### ISSUE TITLE
[Python] Question on the type of Fable.FieldGet

### ISSUE THREAD

**Issue by dbrattli** [fable-compiler/Fable#2644]
### Description

 @alfonsogarciacaro After some recent main merges I have a problem with the Python code and `Fable.FieldGet` to decide if it should be translated to 1) `obj.attribute` or 2) `obj[attribute]`. E.g for anonymous records which are translated to Python dictionaries I need to use the latter subscript syntax.

### Repro code

Usually I would be able to detect if the type is `AnonymousRecord` but in this code from Fable.SimpleJson.Python it becomes `GenericParam`:

```fs
testCase "Deserializing anonymous records" <| fun () ->
    let inputJson = "{ \"content\": [1,2,3] }"
    let deserialized = Json.parseNativeAs<{| content: int list |}> inputJson
    let sum = deserialized.content |> List.sum // FYI .content in this line is the problem
    if sum <> 6 then failwith "Sum should be 6"
```

The current code produces a Fable.FieldGet with a generic type for `deserialized.content`:

`Fable.FieldGet: ("content", GenericParam ("t", []))`

This makes it very hard to know if I should use 1) or 2)

The translated Python code looks like this, but it feels fragile to translate all `GenericParams` to 2) used for anonymous records. I guess generic params could be something else as well needing attribute access.

```py
def arrow_10(_unit=None):
    class ObjectExpr9:
        @property
        def GetZero(self):
            return lambda _unit=None: 0

        @property
        def Add(self):
            return lambda x, y: x + y

    if (
        sum_1(
            Convert_fromJson(
                SimpleJson_parseNative('{ "content": [1,2,3] }'),
                create_type_info(anon_record_type(["content", list_type(int32_type)])),
            )["content"],
            ObjectExpr9(),
        )
        != 6
    ):
        raise Exception("Sum should be 6")


test_case("Deserializing anonymous records", arrow_10)
```

Any ideas?

**Comment by ncave** [comment]
@dbrattli The outer expression (`Fable.Get`) has the actual object type, so you can differentiate on that.

**Comment by dbrattli** [comment]
@ncave Thanks for the hint. But there seems to be some ghost here. I've tried to make a minimal repro without any luck.

But for the example above ([full source](https://github.com/Zaid-Ajaj/Fable.SimpleJson.Python/blob/master/src/Program.fs#L130)) it's still `GenericParam ("t", [])` in the outer `Fable.Get`:

```
transformGet: FieldGet ("content", false)
transformGet: GenericParam ("t", [])
Fable.FieldGet: ("content", GenericParam ("t", []))
```

But if I add a `printfn` like this then the types suddenly becomes correct:

```fs
testCase "Deserializing anonymous records" <| fun () ->
    let inputJson = "{ \"content\": [1,2,3] }"
    let deserialized = Json.parseNativeAs<{| content: int list |}> inputJson
    printfn "%A" deserialized
    let sum = deserialized.content |> List.sum
    if sum <> 6 then failwith "Sum should be 6"
```

```
transformGet: FieldGet ("content", false)
transformGet: AnonymousRecordType ([|"content"|], [List (Number (Int32, None))])
Fable.FieldGet: ("content", AnonymousRecordType ([|"content"|], [List (Number (Int32, None))]))
```

So when compiled to two steps the type is right:

```py
deserialized = Convert_fromJson(SimpleJson_parseNative("{ \"content\": [1,2,3] }"), create_type_info(anon_record_type(["content", list_type(int32_type)])))
if sum_1(deserialized["content"], ObjectExpr9()) != 6:
    raise Exception("Sum should be 6")
```

When it gets inlined the type is wrong:
```py
if sum_1(Convert_fromJson(SimpleJson_parseNative("{ \"content\": [1,2,3] }"), create_type_info(anon_record_type(["content", list_type(int32_type)])))["content"], ObjectExpr9()) != 6:
    raise Exception("Sum should be 6")
```

**Comment by alfonsogarciacaro** [comment]
Yes, it's possible that during transformations of the Fable AST some ident references don't get the concrete type. FableTransforms does some ident replacement when applying lambda beta reduction here: https://github.com/fable-compiler/Fable/blob/f7871e4172764ae1b0c0a9d856768d8fbee3bfa7/src/Fable.Transforms/FableTransforms.fs#L99-L107

However, I realized recently this is not enough because there are other expressions containing idents besides `IdentExpr`. I think all of them are covered now when "resolving" the idents when inlining a function in my latest PR ("resolving" here means making sure the ident names don't conflict with the scope where they're inlined): https://github.com/fable-compiler/Fable/blob/f7871e4172764ae1b0c0a9d856768d8fbee3bfa7/src/Fable.Transforms/FSharp2Fable.fs#L1488-L1495

Are you using latest code from `beyond`? We can try to make sure every transform correctly updates all ident references. But if there are still gaps maybe we can just add a flag to say how the member must be accessed. I also need to add a flag to functions to mark when I don't want them to be compiled as arrow functions (this is necessary for the decorators, because I cannot use `apply(this)` with arrow functions).

**Comment by alfonsogarciacaro** [comment]
Sorry, the comment above is not entirely right. The function in FableTransforms should be fine because we only need to transform ident references not ident declarations. But I still think the problem comes from a transformation that's not correctly propagated. I'll try to reproduce your sample.

**Comment by dbrattli** [comment]
@alfonsogarciacaro Thanks. This was btw my repro attempt, but somehow the types was detected correctly.

```fs
module QuickTest

open System

type Json =
| JNumber of float
| JString of string
| JBool of bool
| JNull
| JArray of Json list
| JObject of Map<string, Json>

[<AutoOpenAttribute>]
module ConverterExtensions =
    let fromJson<'T> json  =
        unbox<'T> json

    type Json with

        static member inline parseNativeAs<'T> input : 'T =
            let a = Activator.CreateInstance(typeof<'T>, [])
            fromJson<'T> a

let test () =
    let deserialized = Json.parseNativeAs<{| content: int list |}>("")
    let sum = deserialized.content |> List.sum
    sum
```

**Comment by ncave** [comment]
@dbrattli Is this specific to `Json.parseNativeAs`, or can it be reproduced without it somehow?

**Comment by dbrattli** [comment]
@ncave I have not been able to reproduce so it might somehow be specific to [`Json.parseNativeAs`](https://github.com/Zaid-Ajaj/Fable.SimpleJson.Python/blob/master/src/Json.Converter.fs#L833). Let's check with @Zaid-Ajaj that have written the code if he knows what could be going on?

FYI: The problem is that Fable somehow reports the type of `deserialized` as ` GenericParam ("t", [])` and not `AnonymousRecordType ([|"content"|], [List (Number (Int32, None))])`

**Comment by alfonsogarciacaro** [comment]
@dbrattli Thanks for sharing the generated code! It looks like the beta reduction has removed `deserialized` and `.content` access directly the inline expr, but the type may not be updated. Does the problem get solved if you add this line [here](https://github.com/fable-compiler/Fable/blob/c223454a3722ea342c773e347c43aefedbf1fa54/src/Fable.Transforms/FSharp2Fable.Util.fs#L1958-L1959)?

![image](https://user-images.githubusercontent.com/8275461/145309527-a31d9b23-2910-4299-bcf9-7a6b948209f4.png)

**Comment by dbrattli** [comment]
@alfonsogarciacaro The fix is working 🚀 🎉 Thank you so much for taking you time with this issue, and great to see again that we can find bugs with one language using a specific library (Fable.SimpleJson.Python) that could have affected other languages as well down the line. So the fix you made will benefit us all!

**Comment by dbrattli** [comment]
Fixed by #2651
